### PR TITLE
fix: recipient

### PR DESCRIPTION
**Pull request by ChefJerry**
### Summary
🐝🐛🛣️

This pull request improves the universal router SDK for PancakeSwap by fixing a bug in the swap logic, adding support for native currencies, and updating the test fixtures and utils. It affects the files `pancakeswap.ts`, `address.ts`, `buildTrade.ts`, and `tokens.ts` in the `packages/universal-router-sdk` directory.

> _`WBNB` mocks BSC_
> _Fixing swaps and routes with_
> _Native currencies_

### Walkthrough
* Fix bug in `PancakeSwapTrade` class that caused incorrect calculation of minimum output amount for exact output trades ([link](https://github.com/pancakeswap/pancake-frontend/pull/8234/files?diff=unified&w=0#diff-566cd8974b99b34b5fdc1c4a109c3da8754c475cba38788149605d459e291b05L90-R90))
* Move logic for unwrapping WETH to user outside of exact output trade condition, since it applies to both exact input and exact output trades ([link](https://github.com/pancakeswap/pancake-frontend/pull/8234/files?diff=unified&w=0#diff-566cd8974b99b34b5fdc1c4a109c3da8754c475cba38788149605d459e291b05L104-R109))
* Add validation checks for `options.recipient` parameter in `addV2Swap`, `addV3Swap`, and `addMixedSwap` functions, since it is required for swap execution ([link](https://github.com/pancakeswap/pancake-frontend/pull/8234/files?diff=unified&w=0#diff-566cd8974b99b34b5fdc1c4a109c3da8754c475cba38788149605d459e291b05L129-R132), [link](https://github.com/pancakeswap/pancake-frontend/pull/8234/files?diff=unified&w=0#diff-566cd8974b99b34b5fdc1c4a109c3da8754c475cba38788149605d459e291b05L171-R172), [link](https://github.com/pancakeswap/pancake-frontend/pull/8234/files?diff=unified&w=0#diff-566cd8974b99b34b5fdc1c4a109c3da8754c475cba38788149605d459e291b05L220-R219))
* Simplify logic for assigning `recipient` variable by using ternary operator instead of nested conditional expression in `addV2Swap`, `addV3Swap`, and `addMixedSwap` functions ([link](https://github.com/pancakeswap/pancake-frontend/pull/8234/files?diff=unified&w=0#diff-566cd8974b99b34b5fdc1c4a109c3da8754c475cba38788149605d459e291b05L129-R132), [link](https://github.com/pancakeswap/pancake-frontend/pull/8234/files?diff=unified&w=0#diff-566cd8974b99b34b5fdc1c4a109c3da8754c475cba38788149605d459e291b05L171-R172), [link](https://github.com/pancakeswap/pancake-frontend/pull/8234/files?diff=unified&w=0#diff-566cd8974b99b34b5fdc1c4a109c3da8754c475cba38788149605d459e291b05L220-R219))
* Import `WBNB` token from `constants/tokens` module and add it to test fixtures for tokens, pairs, and pools on BSC chain ([link](https://github.com/pancakeswap/pancake-frontend/pull/8234/files?diff=unified&w=0#diff-6cb37225df98ceba408ea84bfd04640ea21da81bf981acef444167b03b3b2a37L19-R19), [link](https://github.com/pancakeswap/pancake-frontend/pull/8234/files?diff=unified&w=0#diff-6cb37225df98ceba408ea84bfd04640ea21da81bf981acef444167b03b3b2a37R29), [link](https://github.com/pancakeswap/pancake-frontend/pull/8234/files?diff=unified&w=0#diff-589441c6c966eeae82f2aefa682073c04322c90cadba59fafa1b517860716e45R61-R65))
* Add `WBNB_USDC_V2` pair and `WETH_USDC_V3_MEDIUM_ADDRESS`, `WETH_USDC_V3_LOW`, `WBNB_USDC_V3_MEDIUM`, and `USDC_USDT_V3_LOW_ADDRESS` pools to test fixtures for mixed route trades on BSC and ETH chains ([link](https://github.com/pancakeswap/pancake-frontend/pull/8234/files?diff=unified&w=0#diff-6cb37225df98ceba408ea84bfd04640ea21da81bf981acef444167b03b3b2a37L172-R177), [link](https://github.com/pancakeswap/pancake-frontend/pull/8234/files?diff=unified&w=0#diff-6cb37225df98ceba408ea84bfd04640ea21da81bf981acef444167b03b3b2a37L186-R214))
* Import `ERC20Token` and `Native` types from `@pancakeswap/sdk` module and add `isOutputNative` parameter to `buildMixedRouteTrade` function to handle cases where input or output currency is native or wrapped ([link](https://github.com/pancakeswap/pancake-frontend/pull/8234/files?diff=unified&w=0#diff-d0d10c2a8d3ef71b7a02ca16c25a76a531209e516380995e7ec4ca900bf0f092L1-R1), [link](https://github.com/pancakeswap/pancake-frontend/pull/8234/files?diff=unified&w=0#diff-d0d10c2a8d3ef71b7a02ca16c25a76a531209e516380995e7ec4ca900bf0f092R103))
* Change type of `amounts` array and `outputAmount` variable from `CurrencyAmount<Token>[]` and `CurrencyAmount<Token>` to `CurrencyAmount<Currency>[]` and `CurrencyAmount<Currency>` respectively, since they can contain both wrapped and native currencies ([link](https://github.com/pancakeswap/pancake-frontend/pull/8234/files?diff=unified&w=0#diff-d0d10c2a8d3ef71b7a02ca16c25a76a531209e516380995e7ec4ca900bf0f092L112-R113), [link](https://github.com/pancakeswap/pancake-frontend/pull/8234/files?diff=unified&w=0#diff-d0d10c2a8d3ef71b7a02ca16c25a76a531209e516380995e7ec4ca900bf0f092L118-R122))
* Cast `input` as `CurrencyAmount<ERC20Token>` when calling `getOutputAmount` method on pool, since it expects an ERC20 token as input ([link](https://github.com/pancakeswap/pancake-frontend/pull/8234/files?diff=unified&w=0#diff-d0d10c2a8d3ef71b7a02ca16c25a76a531209e516380995e7ec4ca900bf0f092L118-R122))
* Add logic to unwrap output currency from wrapped token if `isOutputNative` is true and return either native or wrapped output amount depending on parameter ([link](https://github.com/pancakeswap/pancake-frontend/pull/8234/files?diff=unified&w=0#diff-d0d10c2a8d3ef71b7a02ca16c25a76a531209e516380995e7ec4ca900bf0f092L140-R150))

### CODE REVIEW

File: packages/universal-router-sdk/src/entities/protocols/pancakeswap.ts [hunk 1]
```diff
@@ -126,11 +125,10 @@ function addV2Swap(
   // V2 trade should have only one route
   const route = trade.routes[0]
   const path = route.path.map((token) => token.wrapped.address)
-  const recipient = routerMustCustody
-    ? ROUTER_AS_RECIPIENT
-    : typeof options.recipient === 'undefined'
-    ? SENDER_AS_RECIPIENT
-    : validateAndParseAddress(options.recipient)
+
+  if (!options.recipient) throw new Error('recipient is required')
```
Comment (chefjackson) [inline packages/universal-router-sdk/src/entities/protocols/pancakeswap.ts:129]: I think we should accept undefined for recipient? By default should send the fund back to the caller?

Comment (chefjackson) [reply packages/universal-router-sdk/src/entities/protocols/pancakeswap.ts:129]: we better either it as required by extending the existing type `PancakeSwapOptions`, or we allow it to be undefined and use `SENDER_AS_RECIPIENT` by default

Comment (ChefJerry) [reply packages/universal-router-sdk/src/entities/protocols/pancakeswap.ts:129]: default value is assigned by outside

https://github.com/pancakeswap/pancake-frontend/blob/ebf08dcd8234646f7c6488297178cc3afc049d2e/packages/universal-router-sdk/src/entities/protocols/pancakeswap.ts#L52

File: packages/universal-router-sdk/test/utils/buildTrade.ts [hunk 2]
```diff
@@ -137,12 +138,16 @@ export const buildMixedRouteTrade = async <
 
   // mixed Router support exactIn only
   const inputAmount = amount
-  const outputAmount = amounts[amounts.length - 1]
-
+  let outputAmount = amounts[amounts.length - 1]
+  const nativeCurrency = Native.onChain(outputAmount.currency.chainId)
+  // is wrapped token
+  if (outputAmount.currency.wrapped.address === nativeCurrency.wrapped.address && isOutputNative) {
```
Comment (chef-jojo) [inline packages/universal-router-sdk/test/utils/buildTrade.ts:144]: Can we use currency.equal?

**Review by chefjackson — APPROVED** [review]

**Pull request merged**